### ISSUE TITLE
Performance counters

### ISSUE THREAD

**Issue by patriksvensson** [spectresystems/jarvis#3]
To give some insight into what Jarvis is doing.

**Comment by pardahlman** [comment]
I think this is an excellent idea 👌

Any though on how this would be implemented? Serilog has (at least) two different performance libraries: [serilog-metrics](https://github.com/serilog-metrics/serilog-metrics) and [serilog-timings](https://github.com/nblumhardt/serilog-timings). However, Jarvis uses the `IJarvisLog`, so it can not be used right off 😕 

Personally, I would root for using the Serilog `ILog` instead of the `IJarvisLog` and take a dependency to any of the performance libraries. (As a side not, `DebugLog` could be replaced by [debug sink](https://github.com/serilog/serilog-sinks-debug) and the `LogDecorator` with [`LogContext`](https://github.com/serilog/serilog/wiki/Enrichment#the-logcontext)).

Thoughts?

In the future we might want to do something like assigning a query id for each query and enrich log entries with that id (helpful when multiple async queries many providers etc). This can be done with Serilog, but would need some sort of additional code to work with the `IJarvisLog`.

**Comment by patriksvensson** [comment]
@pardahlman I was more thinking along the lines of adding actual [Windows performance counters](https://msdn.microsoft.com/en-us/library/windows/desktop/aa373083(v=vs.85).aspx) to Jarvis that could be ebabled with a pre-processor directive when running performance diagnostics.

**Comment by pardahlman** [comment]
Aah, I see. Never mind that, then 😄

**Comment by patriksvensson** [comment]
@pardahlman I think your suggestions are valid though regardless of performance counters.

Would prefer to have some sort of abstraction around the logger though since I don't want third party addins in the future to have to take a dependency on the libraries that Jarvis use (this makes it difficult to update the libraries if there are ABI breakage). I'm going to add some high level abstraction around the containers as well for registering dependencies [as we've done in Cake](https://github.com/cake-build/cake/tree/develop/src/Cake.Core/Composition).

I like the idea of assigning a query id and enrich log entries with this!

**Comment by pardahlman** [comment]
Good point, I agree with you 👍 . If you want to, I can look at the logging and perhaps:

* Find a way to enrich log entries with properties (and use it to enrich with query id)
* Find a way to add performance logging, I'm thinking perhaps timing around each query execution and perhaps for each query provider?
